### PR TITLE
[5.2] SEF: Enforcing correct SEF URL

### PR DESCRIPTION
**Pull request by Hackwar**
### Summary of Changes
Joomla has been improving its SEO performance constantly and one issue which is still open is, that you can reach the same page via different URLs. This means that you can reach the same article both via the wanted URL, but also via the `/component/content/article/<id>` URL or even via `index.php?option=com_content&view=article&catid=<catid>&id=<id>`. 

This PR adds a new feature which redirects URLs with query parameters to their SEF variant. The router parses the URL and if the original URL contained query parameters, it tries to create a new URL from the information it parsed and if that new URL does not contain any query parameters anymore, it does a redirect to that URL. This will remove a lot of the double URLs for one content item. This only happens for GET requests.

It should be noted, that this does NOT remove every case of duplicate content. For example additional URL parameters are still kept.

I'd like to thank [ithelps Digital](https://www.ithelps-digital.com/) for sponsoring this feature.

### Testing Instructions
1. Apply the patch
2. Use some content to test this with, for example the testing sample data.
3. Go to `/park-blog/first-blog-post` in your site when you use the testing sample data
4. Add a `?id=18` to the URL and see that you actually get the "Second Blog Post" displayed instead of the first one.
5. Enable the option in the SEF plugin.
6. Reload the frontend page and see that you are redirected to `/park-blog/second-blog-post`.

### Link to documentations
Please select:
- [X] Documentation link for docs.joomla.org: https://docs.joomla.org/Search_Engine_Friendly_URLs
- [ ] No documentation changes for docs.joomla.org needed

- [ ] Pull Request link for manual.joomla.org: <link>
- [ ] No documentation changes for manual.joomla.org needed

### CODE REVIEW

**Comment by bembelimen** [comment]
I see what you want to achieve here, but:
- how can a new feature not require documentation (especially such a critical?)
- the whole feature sounds like it's solved when we implement canonical tags

**Comment by Hackwar** [comment]
Sorry about the documentation. It was late yesterday and I will create documentation for the whole SEF thing in the next 48 hours.

Regarding the canonical tag I would disagree. To me, the canonical tag is just a bandaid to get around using the right URL and thus using the right URL in the output is the first step (which we do in Joomla anyway). The second step would then be to redirect wrong URLs to the right ones instead of keeping them and slapping a canonical on there.

**Comment by ceford** [comment]
I have tested this item :white_check_mark: successfully on 6ae36cb3509f4b4bf8ff2f3c323ca257401c7cf6<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/42854">issues.joomla.org/tracker/joomla-cms/42854</a>.</sub>

**Comment by nielsnuebel** [comment]
I have tested this item :white_check_mark: successfully on 6ae36cb3509f4b4bf8ff2f3c323ca257401c7cf6<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/42854">issues.joomla.org/tracker/joomla-cms/42854</a>.</sub>

**Comment by richard67** [comment]
RTC<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/42854">issues.joomla.org/tracker/joomla-cms/42854</a>.</sub>

**Comment by Hackwar** [comment]
Link to documentation has been added.

**Comment by SniperSister** [comment]
I've played around with this PR using the "testing" dataset in Joomla 5.1.x.

I've run into an edge case:
* The testing dataset has an article that isn't accessible via it's own menu item but only via a category list, it's article ID 1, "
Administrator Components": index.php/content-component/article-category-list/extensions/components/administrator-components
* If I now open the "Archive Module" menu item (/index.php/content-component/archive-module) and add an extra query parameter to override the article id with the ID 1 (/index.php/content-component/archive-module?id=1) I'm receiving a redirect to this URL: /index.php/content-component/archive-module?view=article&id=1, whereas my expectation would have been a redirect to the category list URL mentioned above.

I'm aware that this is not quirk of the current PR but of the router in general. However, the current PR might increase the effects.

That's why I'm asking myself if we should limit the redirection to "clean" URLs without any leftover query parameters.

Thoughts? @Hackwar

**Comment by Hackwar** [comment]
I agree, this sounds good. I'll add this to the PR.

**Comment by Hackwar** [comment]
I encountered additional issues and will convert this back to draft first.

**Comment by richard67** [comment]
Back to pending.<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/42854">issues.joomla.org/tracker/joomla-cms/42854</a>.</sub>

**Comment by joeforjoomla** [comment]
What about a generic fetch/xhr request to index.php?option=com_xxx&...
Do you mean that with this option enabled it will no longer be possible to perform a GET request to a raw Joomla URL? 
This would be too much generic.

**Comment by SniperSister** [comment]
> What about a generic fetch/xhr request to index.php?option=com_xxx&...

Fetch / XHR requests do follow redirects automatically.

**Comment by miguelvazquez** [comment]
it works, perfect<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/42854">issues.joomla.org/tracker/joomla-cms/42854</a>.</sub>

File: plugins/system/sef/sef.xml [hunk 1]
```diff
@@ -31,6 +31,19 @@
...
 
+				<field
+					name="enforcesef"
+					type="list"
+					label="PLG_SEF_ENFORCESEF_LABEL"
+					description="PLG_SEF_ENFORCESEF_DESCRIPTION"
+					layout="joomla.form.field.radio.switcher"
+					default="0"
+					filter="boolean"
+				>
```
Comment (Quy) [inline plugins/system/sef/sef.xml]:
```suggestion
					type="radio"
					label="PLG_SEF_ENFORCESEF_LABEL"
					description="PLG_SEF_ENFORCESEF_DESCRIPTION"
					layout="joomla.form.field.radio.switcher"
					default="0"
					filter="boolean"
					>
```

File: administrator/language/en-GB/plg_system_sef.ini [hunk 2]
```diff
@@ -5,6 +5,8 @@
 
 PLG_SEF_DOMAIN_DESCRIPTION="If your site can be accessed through more than one domain enter the preferred (sometimes referred to as canonical) domain here. <br><strong>Note:</strong> https://example.com and https://www.example.com are different domains."
 PLG_SEF_DOMAIN_LABEL="Site Domain"
+PLG_SEF_ENFORCESEF_DESCRIPTION="When the URL contains non-SEF parts or the page is normally reachable via a different URL, this option forces the browser to redirect to the correct URL."
+PLG_SEF_ENFORCESEF_LABEL="Disallow non-SEF URLs"
```
Comment (SniperSister) [inline administrator/language/en-GB/plg_system_sef.ini]: In the light of the usage of this parameter in #42989 I would suggest to rename it to "Enforce strict SEF URLs"

Comment (Hackwar) [reply administrator/language/en-GB/plg_system_sef.ini]: I added a new option "Strict Routing" in #43432, which this PR depends on and refactored this PR accordingly.

**Comment by Hackwar** [comment]
I refactored the PR to use the new option from #43432 to reduce the number of options we add here.

**Comment by robbiejackson** [comment]
I have tested this item :white_check_mark: successfully on 09f8ba345508d55ca324fa01218c5eef2f191869<br /><br />I tested this using configuration which resulted in a /component/content URL. Then I enabled a menuitem which matched this URL and the code successfully did a 301 redirect to the SEF URL for that menuitem. I repeated this for several cases and they all worked ok.

I also set up a site module which was a list of articles in a given category. The URLs of the articles were based on a category list menuitem, and when I clicked on one of the links it routed me correctly to the article.

I then set up a menuitem pointing directly at the single article and reloaded the site page. It didn't redirect to the new menuitem (as there weren't any query parameters on the URL). But when the module displayed the links it utilised the menuitem for that single article. <hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/42854">issues.joomla.org/tracker/joomla-cms/42854</a>.</sub>

**Comment by Hackwar** [comment]
RTC<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/42854">issues.joomla.org/tracker/joomla-cms/42854</a>.</sub>

**Comment by pe7er** [comment]
Thanks @Hackwar !

**Pull request merged**